src/Type.h

```diff
@@ -511,10 +511,22 @@ class FuncType final : public Type
 	/**
 	 * A single lambda "capture" (outer variable used in a lambda's body).
 	 */
-	struct Capture
+	class Capture
 		{
+	public:
+		Capture(detail::IDPtr _id, bool _deep_copy);
+
+		auto& Id() const { return id; }
+		bool IsDeepCopy() const { return deep_copy; }
+		bool IsManaged() const { return is_managed; }
+
+		// For script optimization:
+		void SetID(detail::IDPtr new_id) { id = std::move(new_id); }
+
+	private:
```

Comment:
Yeah that was never meant to be a public API. I'm really un-jazzed about how hard C++ makes it to cleanly separate out namespaces (or maybe it doesn't, but our style of using them does?). I never even think to check whether I'm adding stuff to a non-`detail` namespace given how often I'm living in the `detail` world with all this AST finery.

Here, I can't conceive of a situation where a Zeek developer not working on the internals would ever want to create a `Capture` object, since their only use is by the internals. (1) How would I have known that I'm creating something in the non-`detail` space? (2) If I _wanted_ it to be in the `detail` space (or otherwise hidden from developers), how should I have done that?

Reply:
> Yeah that was never meant to be a public API

Seems pretty reasonable to provide a stable way for external users introspect the list of captures of a `FuncType` for lambdas.

Hypothetical example: A plugin providing a built-in function taking a lambda and pretty printing Params(), Yield() and CaptureList(). Separately, and I suspect script_opt does, external plugins could compile or otherwise analyze code within lambdas and would probably require access to the capture list to achieve that goal.

> Here, I can't conceive of a situation where a Zeek developer not working on the internals would ever want to create a Capture object,

Creating the Capture object is one aspect, the bigger issue is hiding the fields and replacing with accessors without a transition period.

>  (1) How would I have known that I'm creating something in the non-detail space?

All Type classes are in the to-level `zeek` namespace. Adding a nested class into one of these makes it part of `zeek`. When I review I either scroll-up or hover over the identifiers to see what the namespace is.

> (2) If I wanted it to be in the detail space (or otherwise hidden from developers), how should I have done that?

Not sure there's a good answer here. Putting the `Capture` struct in  `zeek::detail::` seems not right when the only method returning those is in `zeek` namespace and public.

Reply:
So what's the way forward with this one?  I'm fine adding a default constructor that's marked as deprecated. That though doesn't address the need to deprecate direct access to the individual member variables. Can those just be annotated with `[[deprecated]]` without leading to complaints generated by the class's own methods?

Reply:
> Can those just be annotated with [[deprecated]] without leading to complaints generated by the class's own methods?

Haven't tried, but `[[deprecated]]` on the fields and then bracketing with ignoring deprecated-declarations should work.
```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"

constructor

accessors

#pragma GCC diagnostic pop
```

It's not great, but it does give a consumer a nicer transition phase.